### ISSUE TITLE
2019.07.15

### ISSUE THREAD

**Issue by leeEunJooo** [leeEunJooo/machine-learning#1]
머신러닝
기본 내용
mport pandas as pd
 
data = [1, 3, 5, 7, 9]
s = pd.Series(data)
https://stickie.tistory.com/43?category=746224
https://datascienceschool.net/view-notebook/293ece8b0d124fbaa4d4d52bb8f1cb42/

SKLEARN
from sklearn import svm
#scikit-learn 에서 svm(support vector machine을 사용)
xor_data = [
    [0, 0, 0],
    [0, 1, 1],
    [1, 0, 1],
    [1, 1, 0]
] 
CLF
후에 쓸 clf.fit()의 매개변수로 학습할 데이터 배열, 레이블 배열을 전달해야 하기 때문에 이 데이터를 쪼갤것이다.
각 배열 당 앞의 두 값을 data라는 리스트에, 제일 뒤의 값을 label(정답의 느낌)에 넣어야 한다.
data =[]
label =[]
for row in xor_data:
    p = row[0]
    q = row[1]
    r = row[2]
    data.append([p, q])
    label.append(r)

CLF
svm중에 clf를 가지고 옴
clf = svm.SVC()
clf.fit(data, label)

predicting = clf.predict(data)
print("예측결과:", predicting)

ok = 0
total = 0 
for idx, answer in enumerate(label):
    p=predicting[idx]
    if p == answer:
        ok +=1
    total += 1
print('정답률:', ok, "/", total, "=", ok/total)

Sklearn 사용해서 더 간단하게 하기

import pandas as pd
from sklearn import svm, metrics 
xor_input= [
    [0, 0, 0],
    [0, 1, 1],
    [1, 0, 1],
    [1, 1, 0]
]

xor_df = pd.DataFrame(xor_input)
xor_data = xor_df.loc[:, 0:1]
xor_label = xor_df.loc[:, 2]

clf = svm.SVC()
clf.fit(xor_data, xor_label)
pre = clf.predict(xor_data)

print(pre)

ac_score = metrics.accuracy_score(xor_label, pre)
print("정답률=", ac_score)

https://doorbw.tistory.com/172

https://github.com/pydata/pandas/blob/master/pandas/tests/data/iris.csv\\

csv는 콤마로되어있음

테이블 저장

from sklearn import svm, metrics
import random, re

csv = []
with open('iris.csv', 'r', encoding='utf-8') as fp:
   for line in fp:
      line = line.strip() #줄바꿈 제거
      cols = line.split(',') #쉼표로 자르기

      fn = lambda n : float(n) if re.match(r'^[0-9\.]+$', n) else n #각 셀의 데이터가 현재 숫자처럼 생긴 String로 되어있기 때문에, 
        #이게 숫자를 의미하는지 확인한 후에 float()에 넣어서 진짜 숫자로 바꿈. 
      cols = list(map(fn, cols)) #python에서 map은 처음에 오는 매개변수(함수)에 다음에 오는 매개변수(여기서 cols)를 넣어 실행하는 것
        # list로 감싸서 이를 리스트화해서 다시 cols에 넣음. 
      csv.append(cols)

del csv[0]

random.shuffle(csv) #순서 섞기

total_len = len(csv)
train_len = int(total_len * 2 / 3) #통상적으로 전체 데이터의 2/3 양을 train_data로 만듦
train_data = []
train_label = []
test_data = []
test_label = []

for i in range(total_len): 
   data = csv[i][0:4]
   label = csv[i][4]

   if i < train_len:
      train_data.append(data)
      train_label.append(label)
   else:
      test_data.append(data)
      test_label.append(label)

clf = svm.SVC()
clf.fit(train_data, train_label)
pre = clf.predict(test_data)

ac_score = metrics.accuracy_score(test_label, pre)
print("정답률 = ", ac_score)

다른 더 짧은 방법
import pandas as pd
from sklearn import svm, metrics
from sklearn.model_selection import train_test_split

csv = pd.read_csv("iris.csv")
data = csv[["SepalLength","SepalWidth","PetalLength","PetalWidth"]]
label = csv["Name"]

train_data, test_data, train_label, test_label = train_test_split(data,label)

clf = svm.SVC()
clf.fit(train_data,train_label)
results=clf.predict(test_data)

score = metrics.accuracy_score(results,test_label)

print("정답률:",score)
